workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/selection/VaultWatcherService.kt

```diff
@@ -0,0 +1,148 @@
+package com.r3.corda.sdk.token.workflow.selection
+
+import com.r3.corda.sdk.token.contracts.states.FungibleToken
+import com.r3.corda.sdk.token.contracts.types.IssuedTokenType
+import com.r3.corda.sdk.token.contracts.types.TokenType
+import com.r3.corda.sdk.token.workflow.selection.VaultWatcherService.Companion.LOG
+import net.corda.core.contracts.Amount
+import net.corda.core.contracts.StateAndRef
+import net.corda.core.node.AppServiceHub
+import net.corda.core.node.services.CordaService
+import net.corda.core.node.services.Vault
+import net.corda.core.node.services.vault.DEFAULT_PAGE_NUM
+import net.corda.core.node.services.vault.PageSpecification
+import net.corda.core.utilities.contextLogger
+import rx.Observable
+import java.lang.RuntimeException
+import java.security.PublicKey
+import java.time.Duration
+import java.util.concurrent.*
+
+val unlocker: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
+
+@CordaService
+class VaultWatcherService(vaultObserver: VaultObserver? = null) {
+
+    val cache: ConcurrentMap<PublicKey, ConcurrentMap<Class<*>, ConcurrentMap<String, ConcurrentMap<StateAndRef<FungibleToken<TokenType>>, Boolean>>>> = ConcurrentHashMap()
+
+    constructor(appServiceHub: AppServiceHub) : this(getObservableFromAppServiceHub(appServiceHub))
+
+    companion object {
+        val LOG = contextLogger()
+
+        private fun getObservableFromAppServiceHub(appServiceHub: AppServiceHub): VaultObserver {
+            val pageSize = 1000
+            var currentPage = DEFAULT_PAGE_NUM;
+            var (existingStates, observable) = appServiceHub.vaultService.trackBy(FungibleToken::class.java, PageSpecification(pageNumber = currentPage, pageSize = pageSize))
```

Comment:
You might want to replace this with a JPA query directly for `PersistentFungibleToken` to initialise at start up to make things a lot faster.  If that table can hold the fields needed to bucket these, plus the amount, that would make the init sequence much faster.  Loading all of the whole states via the vault will be very expensive.